File: src/com/sleazyweasel/applescriptifier/preferences/MuseControllerPreferences.java
package com.sleazyweasel.applescriptifier.preferences;


import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.prefs.BackingStoreException;
import java.util.prefs.Preferences;

public class MuseControllerPreferences {
    private static final Logger logger = Logger.getLogger(MuseControllerPreferences.class.getName());
    private static final String PANDORA_VETO_KEY = "pianobar.veto";
    private static final String SPOTIFY_ENABLE_KEY = "spotify.enable";
    private static final String MUSECONTROL_ENABLE_KEY = "musecontrol.enable";
    private static final String LAST_STREAMER_KEY = "last.streamer";

    private static final String PANDORA_STREAMER_VALUE = "Pandora";
    private static final String SPOTIFY_STREAMER_VALUE = "Spotify";
    private static final String PREVIOUS_PANDORA_STATION_ID_KEY = "PREVIOUS_PANDORA_STATION_ID_KEY";
    private static final String PREVIOUS_PANDORA_VOLUME_KEY = "PREVIOUS_PANDORA_VOLUME_KEY";
    private static final String PREVIOUS_SPOTIFY_VOLUME_KEY = "PREVIOUS_SPOTIFY_VOLUME_KEY";

    private final Preferences preferences;

    public MuseControllerPreferences(Preferences preferences) {
        this.preferences = preferences;
    }

    public boolean isPandoraEnabled() {
        return !preferences.getBoolean(PANDORA_VETO_KEY, false);
    }

    public void enablePandora(boolean enable) {
        preferences.putBoolean(PANDORA_VETO_KEY, !enable);
    }

    public void enableMuseControl(boolean enable) {
        preferences.putBoolean(MUSECONTROL_ENABLE_KEY, enable);
    }

    public boolean isMuseControlEnabled() {
        return preferences.getBoolean(MUSECONTROL_ENABLE_KEY, true);
    }

    public void save() throws BackingStoreException {
        preferences.sync();
    }

    public boolean wasPandoraTheLastStreamerOpen() {
        String lastStreamer = preferences.get(LAST_STREAMER_KEY, null);
        //at the moment, Pandora is the default to start up with. In future, I should probably pop up a message on first startup of the app, to configure prefs and set this stuff up.
        return lastStreamer == null || lastStreamer.equals(PANDORA_STREAMER_VALUE);
    }

    public boolean wasSpotifyTheLastStreamerOpen() {
        String lastStreamer = preferences.get(LAST_STREAMER_KEY, null);
        return SPOTIFY_STREAMER_VALUE.equals(lastStreamer);
    }

    public void setPandoraAsStreamer() {
        preferences.put(LAST_STREAMER_KEY, PANDORA_STREAMER_VALUE);
    }

    public void enableSpotify(boolean enable) {
        preferences.putBoolean(SPOTIFY_ENABLE_KEY, enable);
    }

    public void setSpotifyAsStreamer() {
        preferences.put(LAST_STREAMER_KEY, SPOTIFY_STREAMER_VALUE);
    }

    public Long getPreviousPandoraStationId() {
        if (keyExists(PREVIOUS_PANDORA_STATION_ID_KEY)) {
            return preferences.getLong(PREVIOUS_PANDORA_STATION_ID_KEY, 0);
        }
        return null;
    }

    public void setPandoraStationId(Long stationId) {
        if (stationId == null) {
            preferences.remove(PREVIOUS_PANDORA_STATION_ID_KEY);
        }
        else {
            preferences.putLong(PREVIOUS_PANDORA_STATION_ID_KEY, stationId);
        }
    }

    public Double getPreviousPandoraVolume() {
        if (keyExists(PREVIOUS_PANDORA_VOLUME_KEY)) {
            return preferences.getDouble(PREVIOUS_PANDORA_VOLUME_KEY, 0);
        }
        return null;
    }

    public void setPandoraVolume(double volume) {
        preferences.putDouble(PREVIOUS_PANDORA_VOLUME_KEY, volume);
    }

    public Float getPreviousSpotifyVolume() {
        return preferences.getFloat(PREVIOUS_SPOTIFY_VOLUME_KEY, 1.0f);
    }

    private boolean keyExists(String keyToCheck) {
        try {
            String[] keys = preferences.keys();
            for (String key : keys) {
                if (keyToCheck.equals(key)) {
                    return true;
                }
            }
            return false;
        } catch (BackingStoreException e) {
            logger.log(Level.WARNING, "Exception caught.", e);
            return false;
        }
    }

}

File: src/com/sleazyweasel/applescriptifier/preferences/MuseControllerPreferences.scala
package com.sleazyweasel.applescriptifier.preferences

import java.util.logging.Level
import java.util.logging.Logger
import java.util.prefs.BackingStoreException
import java.util.prefs.Preferences

object MuseControllerPreferences {
  private final val logger: Logger = Logger.getLogger(classOf[MuseControllerPreferences].getName)
  private final val PANDORA_VETO_KEY: String = "pianobar.veto"
  private final val SPOTIFY_ENABLE_KEY: String = "spotify.enable"
  private final val MUSECONTROL_ENABLE_KEY: String = "musecontrol.enable"
  private final val LAST_STREAMER_KEY: String = "last.streamer"
  private final val PANDORA_STREAMER_VALUE: String = "Pandora"
  private final val SPOTIFY_STREAMER_VALUE: String = "Spotify"
  private final val PREVIOUS_PANDORA_STATION_ID_KEY: String = "PREVIOUS_PANDORA_STATION_ID_KEY"
  private final val PREVIOUS_PANDORA_VOLUME_KEY: String = "PREVIOUS_PANDORA_VOLUME_KEY"
  private final val PREVIOUS_SPOTIFY_VOLUME_KEY: String = "PREVIOUS_SPOTIFY_VOLUME_KEY"
}

class MuseControllerPreferences(val preferences: Preferences) {

  def isPandoraEnabled: Boolean = {
    !preferences.getBoolean(MuseControllerPreferences.PANDORA_VETO_KEY, false)
  }

  def enablePandora(enable: Boolean) {
    preferences.putBoolean(MuseControllerPreferences.PANDORA_VETO_KEY, !enable)
  }

  def enableMuseControl(enable: Boolean) {
    preferences.putBoolean(MuseControllerPreferences.MUSECONTROL_ENABLE_KEY, enable)
  }

  def isMuseControlEnabled: Boolean = {
    preferences.getBoolean(MuseControllerPreferences.MUSECONTROL_ENABLE_KEY, true)
  }

  def save() {
    preferences.sync()
  }

  def wasPandoraTheLastStreamerOpen(): Boolean = {
    val lastStreamer: String = preferences.get(MuseControllerPreferences.LAST_STREAMER_KEY, null)
    lastStreamer == null || (lastStreamer == MuseControllerPreferences.PANDORA_STREAMER_VALUE)
  }

  def wasSpotifyTheLastStreamerOpen(): Boolean = {
    val lastStreamer: String = preferences.get(MuseControllerPreferences.LAST_STREAMER_KEY, null)
    MuseControllerPreferences.SPOTIFY_STREAMER_VALUE == lastStreamer
  }

  def setPandoraAsStreamer() {
    preferences.put(MuseControllerPreferences.LAST_STREAMER_KEY, MuseControllerPreferences.PANDORA_STREAMER_VALUE)
  }

  def enableSpotify(enable: Boolean) {
    preferences.putBoolean(MuseControllerPreferences.SPOTIFY_ENABLE_KEY, enable)
  }

  def setSpotifyAsStreamer() {
    preferences.put(MuseControllerPreferences.LAST_STREAMER_KEY, MuseControllerPreferences.SPOTIFY_STREAMER_VALUE)
  }

  def getPreviousPandoraStationId: Long = {
    if (keyExists(MuseControllerPreferences.PREVIOUS_PANDORA_STATION_ID_KEY)) {
      return preferences.getLong(MuseControllerPreferences.PREVIOUS_PANDORA_STATION_ID_KEY, 0)
    }
    -1
  }

  def setPandoraStationId(stationId: Long) {
    if (stationId == -1) {
      preferences.remove(MuseControllerPreferences.PREVIOUS_PANDORA_STATION_ID_KEY)
    }
    else {
      preferences.putLong(MuseControllerPreferences.PREVIOUS_PANDORA_STATION_ID_KEY, stationId)
    }
  }

  def getPreviousPandoraVolume: Double = {
    if (keyExists(MuseControllerPreferences.PREVIOUS_PANDORA_VOLUME_KEY)) {
      return preferences.getDouble(MuseControllerPreferences.PREVIOUS_PANDORA_VOLUME_KEY, 0)
    }
    -1d
  }

  def setPandoraVolume(volume: Double) {
    preferences.putDouble(MuseControllerPreferences.PREVIOUS_PANDORA_VOLUME_KEY, volume)
  }

  def getPreviousSpotifyVolume: Float = {
    preferences.getFloat(MuseControllerPreferences.PREVIOUS_SPOTIFY_VOLUME_KEY, 1.0f)
  }

  private def keyExists(keyToCheck: String): Boolean = {
    try {
      val keys: Array[String] = preferences.keys
      for (key <- keys) {
        if (keyToCheck == key) {
          return true
        }
      }
      false
    }
    catch {
      case e: BackingStoreException => {
        MuseControllerPreferences.logger.log(Level.WARNING, "Exception caught.", e)
        return false
      }
    }
  }

}